File: kotlin/src/wildling/ParsePattern.kt
package wildling

import java.util.regex.Pattern

typealias Dictionaries = Map<String, List<String>>

internal object ParsePattern {
    private val TOKEN_PARSING_REGEX: Pattern = Pattern.compile(
        "(\\\\[%@$*#&?!-]|[%@$*#&?!-]\\{.*?\\}|[%@$*#&?!-])"
    )
    private val LENGTH_WITH_VARIANTS: Pattern = Pattern.compile(
        "\\{((\\d+)-(\\d+)|(\\d+))\\}"
    )
    private val LENGTH_WITH_STRING: Pattern = Pattern.compile(
        "\\{'(.*)'(?:,(\\d+)-(\\d+))?(?:,(\\d+))?\\}"
    )

    private fun parseLengthWithVariants(part: String, variants: List<String>): TokenOptions {
        val match = LENGTH_WITH_VARIANTS.matcher(part)
        var startLength = 1
        var endLength = 1

        if (match.find()) {
            if (match.group(2) != null) {
                startLength = match.group(2).toInt()
                endLength = match.group(3).toInt()
            } else if (match.group(1) != null) {
                startLength = match.group(1).toInt()
                endLength = startLength
            }
        }

        return TokenOptions(
            variants = variants,
            startLength = startLength,
            endLength = endLength,
            src = part,
        )
    }

    private fun parseLengthWithString(part: String): TokenOptions? {
        val match = LENGTH_WITH_STRING.matcher(part)
        if (!match.find()) {
            return null
        }

        val string = match.group(1) ?: ""
        if (match.group(2) != null && match.group(3) != null) {
            return TokenOptions(
                string = string,
                startLength = match.group(2).toInt(),
                endLength = match.group(3).toInt(),
                src = part,
            )
        }

        if (match.group(4) != null) {
            val length = match.group(4).toInt()
            return TokenOptions(
                string = string,
                startLength = length,
                endLength = length,
                src = part,
            )
        }

        return TokenOptions(
            string = string,
            startLength = 1,
            endLength = 1,
            src = part,
        )
    }

    private fun simpleTokenizer(variantsString: String): (String) -> Token {
        val variants = variantsString.map { it.toString() }
        return { part -> Token(parseLengthWithVariants(part, variants)) }
    }

    private fun dictionaryTokenizer(part: String, dictionaries: Dictionaries): Token {
        var options = parseLengthWithString(part)
        val key = options?.string
        if (options == null || (key != null && key.isNotEmpty() && key !in dictionaries)) {
            options = TokenOptions(
                variants = listOf(part),
                startLength = 1,
                endLength = 1,
                src = part,
            )
        } else {
            options.variants = dictionaries[key ?: ""] ?: emptyList()
        }
        return Token(options)
    }

    private fun wordsTokenizer(part: String): Token {
        var options = parseLengthWithString(part)
        if (options == null) {
            options = TokenOptions(
                variants = listOf(part),
                startLength = 1,
                endLength = 1,
                src = part,
            )
        } else {
            val variants = mutableListOf<String>()
            var workString = options.string ?: ""
            var index = 0
            while (index < workString.length) {
                if (index + 1 < workString.length &&
                    workString[index] == '\\' &&
                    workString[index + 1] == ','
                ) {
                    index += 2
                } else if (workString[index] == ',') {
                    variants.add(workString.substring(0, index))
                    workString = workString.substring(index + 1)
                    index = 0
                } else {
                    index += 1
                }
            }
            variants.add(workString)
            options.variants = variants.map { it.replace("\\,", ",") }
        }
        return Token(options)
    }

    private fun partToToken(part: String, dictionaries: Dictionaries): Token {
        val tokenizers = mapOf(
            '#' to simpleTokenizer("0123456789"),
            '@' to simpleTokenizer("abcdefghijklmnopqrstuvwxyz"),
            '*' to simpleTokenizer("abcdefghijklmnopqrstuvwxyz0123456789"),
            '-' to simpleTokenizer(
                "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
            ),
            '!' to simpleTokenizer("ABCDEFGHIJKLMNOPQRSTUVWXYZ"),
            '?' to simpleTokenizer("ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"),
            '&' to simpleTokenizer(
                "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
            ),
            '%' to { p: String -> dictionaryTokenizer(p, dictionaries) },
            '$' to { p: String -> wordsTokenizer(p) },
        )

        val tokenizer = if (part.isNotEmpty()) tokenizers[part[0]] else null
        val isEscaped =
            part.length > 1 && part[0] == '\\' && tokenizers.containsKey(part[1])

        return when {
            tokenizer != null -> tokenizer(part)
            isEscaped -> Token(
                TokenOptions(
                    variants = listOf(part.substring(1)),
                    startLength = 1,
                    endLength = 1,
                    src = part,
                )
            )
            else -> Token(
                TokenOptions(
                    variants = listOf(part),
                    startLength = 1,
                    endLength = 1,
                    src = part,
                )
            )
        }
    }

    /** Split like JS/Python capturing-group split (Java Pattern.split does not). */
    private fun splitKeepingDelimiters(input: String): List<String> {
        val parts = mutableListOf<String>()
        val matcher = TOKEN_PARSING_REGEX.matcher(input)
        var last = 0
        while (matcher.find()) {
            if (matcher.start() > last) {
                val before = input.substring(last, matcher.start())
                if (before.isNotEmpty()) {
                    parts.add(before)
                }
            }
            val token = matcher.group(1)
            if (!token.isNullOrEmpty()) {
                parts.add(token)
            }
            last = matcher.end()
        }
        if (last < input.length) {
            val rest = input.substring(last)
            if (rest.isNotEmpty()) {
                parts.add(rest)
            }
        }
        return parts
    }

    fun parse(inputPattern: String, dictionaries: Dictionaries?): List<Token> {
        val dicts = dictionaries ?: emptyMap()
        return splitKeepingDelimiters(inputPattern).map { partToToken(it, dicts) }
    }
}

File: scala/src/wildling/ParsePattern.scala
package wildling

import java.util.regex.Pattern
import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer

object ParsePattern {
  type Dictionaries = Map[String, Seq[String]]

  private val TokenParsingRegex: Pattern = Pattern.compile(
    "(\\\\[%@$*#&?!-]|[%@$*#&?!-]\\{.*?\\}|[%@$*#&?!-])"
  )
  private val LengthWithVariants: Pattern = Pattern.compile(
    "\\{((\\d+)-(\\d+)|(\\d+))\\}"
  )
  private val LengthWithString: Pattern = Pattern.compile(
    "\\{'(.*)'(?:,(\\d+)-(\\d+))?(?:,(\\d+))?\\}"
  )

  private def parseLengthWithVariants(part: String, variants: Seq[String]): TokenOptions = {
    val matchResult = LengthWithVariants.matcher(part)
    var startLength = 1
    var endLength = 1

    if (matchResult.find()) {
      if (matchResult.group(2) != null) {
        startLength = matchResult.group(2).toInt
        endLength = matchResult.group(3).toInt
      } else if (matchResult.group(1) != null) {
        startLength = matchResult.group(1).toInt
        endLength = startLength
      }
    }

    TokenOptions.of(variants, startLength, endLength, part)
  }

  private def parseLengthWithString(part: String): Option[TokenOptions] = {
    val matchResult = LengthWithString.matcher(part)
    if (!matchResult.find()) return None

    val string = Option(matchResult.group(1)).getOrElse("")
    if (matchResult.group(2) != null && matchResult.group(3) != null) {
      return Some(
        new TokenOptions(
          string = Some(string),
          startLength = Some(matchResult.group(2).toInt),
          endLength = Some(matchResult.group(3).toInt),
          src = Some(part)
        )
      )
    }

    if (matchResult.group(4) != null) {
      val length = matchResult.group(4).toInt
      return Some(
        new TokenOptions(
          string = Some(string),
          startLength = Some(length),
          endLength = Some(length),
          src = Some(part)
        )
      )
    }

    Some(
      new TokenOptions(
        string = Some(string),
        startLength = Some(1),
        endLength = Some(1),
        src = Some(part)
      )
    )
  }

  private def simpleTokenizer(variantsString: String): String => Token = {
    val variants = variantsString.map(_.toString)
    (part: String) => new Token(parseLengthWithVariants(part, variants))
  }

  private def dictionaryTokenizer(part: String, dictionaries: Dictionaries): Token = {
    var options = parseLengthWithString(part)
    val key = options.flatMap(_.string)
    if (
      options.isEmpty ||
      key.exists(k => k.nonEmpty && !dictionaries.contains(k))
    ) {
      options = Some(TokenOptions.of(Seq(part), 1, 1, part))
    } else {
      val opts = options.get
      opts.variants = Some(dictionaries.getOrElse(key.getOrElse(""), Seq.empty))
    }
    new Token(options.get)
  }

  private def wordsTokenizer(part: String): Token = {
    var options = parseLengthWithString(part)
    if (options.isEmpty) {
      options = Some(TokenOptions.of(Seq(part), 1, 1, part))
    } else {
      val opts = options.get
      val variants = ArrayBuffer.empty[String]
      var workString = opts.string.getOrElse("")
      var index = 0
      while (index < workString.length) {
        if (
          index + 1 < workString.length &&
          workString.charAt(index) == '\\' &&
          workString.charAt(index + 1) == ','
        ) {
          index += 2
        } else if (workString.charAt(index) == ',') {
          variants += workString.substring(0, index)
          workString = workString.substring(index + 1)
          index = 0
        } else {
          index += 1
        }
      }
      variants += workString
      opts.variants = Some(variants.map(_.replace("\\,", ",")).toSeq)
    }
    new Token(options.get)
  }

  private def partToToken(part: String, dictionaries: Dictionaries): Token = {
    val tokenizers = mutable.Map[Char, String => Token](
      '#' -> simpleTokenizer("0123456789"),
      '@' -> simpleTokenizer("abcdefghijklmnopqrstuvwxyz"),
      '*' -> simpleTokenizer("abcdefghijklmnopqrstuvwxyz0123456789"),
      '-' -> simpleTokenizer(
        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
      ),
      '!' -> simpleTokenizer("ABCDEFGHIJKLMNOPQRSTUVWXYZ"),
      '?' -> simpleTokenizer("ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"),
      '&' -> simpleTokenizer("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"),
      '%' -> ((p: String) => dictionaryTokenizer(p, dictionaries)),
      '$' -> ((p: String) => wordsTokenizer(p))
    )

    val tokenizer = if (part.nonEmpty) tokenizers.get(part.charAt(0)) else None
    val isEscaped =
      part.length > 1 && part.charAt(0) == '\\' && tokenizers.contains(part.charAt(1))

    if (tokenizer.isDefined) {
      tokenizer.get(part)
    } else if (isEscaped) {
      new Token(TokenOptions.of(Seq(part.substring(1)), 1, 1, part))
    } else {
      new Token(TokenOptions.of(Seq(part), 1, 1, part))
    }
  }

  /** Split like JS/Python capturing-group split. */
  private def splitKeepingDelimiters(input: String): Seq[String] = {
    val parts = ArrayBuffer.empty[String]
    val matcher = TokenParsingRegex.matcher(input)
    var last = 0
    while (matcher.find()) {
      if (matcher.start() > last) {
        val before = input.substring(last, matcher.start())
        if (before.nonEmpty) parts += before
      }
      val token = matcher.group(1)
      if (token != null && token.nonEmpty) parts += token
      last = matcher.end()
    }
    if (last < input.length) {
      val rest = input.substring(last)
      if (rest.nonEmpty) parts += rest
    }
    parts.toSeq
  }

  def parse(inputPattern: String, dictionaries: Dictionaries): Seq[Token] = {
    val dicts = if (dictionaries == null) Map.empty[String, Seq[String]] else dictionaries
    splitKeepingDelimiters(inputPattern).map(partToToken(_, dicts))
  }
}
